thrift/thrift.bzl

```diff
@@ -3,37 +3,41 @@
 _thrift_filetype = FileType([".thrift"])
 
 def _thrift_library_impl(ctx):
+  #TODO MAKE THIS BE A SCALA IMPLEMENTATION, AND USE THE ABSOLUTE PREFIX!
+  prefix = ctx.attr.absolute_prefix
+  jarcmd = "{jar} cMf {out} -C {out}_tmp ."
+  if prefix != '':
+    jarcmd = "{{jar}} cMf {{out}} -C {{out}}_tmp/{prefix} .".format(prefix=prefix)
+
   _valid_thrift_deps(ctx.attr.deps)
   # We move the files and touch them so that the output file is a purely deterministic
   # product of the _content_ of the inputs
-  #TODO is rsync an acceptable dependency here?
   cmd = """
 rm -rf {out}_tmp
 mkdir -p {out}_tmp
 {jar} cMf {out}_tmp/tmp.jar $@
-unzip -o {out}_tmp/tmp.jar -d {out}_tmp >/dev/null
+unzip -o {out}_tmp/tmp.jar -d {out}_tmp 2>/dev/null
 rm -rf {out}_tmp/tmp.jar
 find {out}_tmp -exec touch -t 198001010000 {{}} \;
-{jar} cMf {out} -C {out}_tmp .
-rm -rf {out}_tmp""".format(out=ctx.outputs.libarchive.path,
-                           jar=ctx.file._jar.path)
+""" + jarcmd + """
+rm -rf {out}_tmp"""
+
+  cmd = cmd.format(out=ctx.outputs.libarchive.path,
+                   jar=ctx.file._jar.path)
   ctx.action(
     inputs = ctx.files.srcs + ctx.files._jar + ctx.files._jdk,
     outputs = [ctx.outputs.libarchive],
     command = cmd,
     progress_message = "making thrift archive %s" % ctx.label,
     arguments = [f.path for f in ctx.files.srcs],
   )
-  transitive_archives = _collect_thrift_tars(ctx.attr.deps)
-  transitive_archives += [ctx.outputs.libarchive]
 
   transitive_srcs = _collect_thrift_srcs(ctx.attr.deps)
-  transitive_srcs += ctx.files.srcs
+  transitive_srcs += [ctx.outputs.libarchive]
```

Comment:
so one of the changes that I made is that we actually use the archive produced by thrift_library. this made everything much more consistent, and also helped the actual code gen, because instead of wrangling sources, I just have archives that I could unzip wherever I wanted. also, it felt more in line with the bazel philosophy of depending on the outputs of previous targets, instead of just sort of grabbing the file tree.